// cal/lib/middleware.js
'use strict';

var nsutils = require('./nsutils');
var threads = require('./threads');
var constants = require('./constants');
var tryRequire = require('try-require');
var onFinished = require('on-finished');
var correlationId = require('./correlationId');
var threadCache = require('./thread-cache');

module.exports = function middleware(options) {
    options = options || {};

    var enableBuffering = (typeof options.enableBuffering === 'boolean') ? options.enableBuffering : false;
    var enableNestedCal = options.enableNestedCal === undefined ? false : options.enableNestedCal;
    var bufferFlushDelay = options.bufferFlushDelay || 0;

    return function cal(req, res, next) {
        nsutils.bindEmitter(req);
        nsutils.bindEmitter(res);

        var threadId = threads.getNextThreadId();

        var context = {
            correlationId: correlationId(req),
            bufferFlushDelay: bufferFlushDelay,
            isBufferingEnabled: enableBuffering,
            isNestedCal: enableNestedCal
        };
        threadCache.set(threadId, context);

        nsutils.run(function () {
            nsutils.set(constants.TX, null);
            nsutils.set(constants.THREAD_ID, threadId);
            req.threadId = threadId;    //For backward compatibility
            nsutils.set('req', req);
            nsutils.set('correlationId', context.correlationId);
            // reset counter
            nsutils.set('nextId', null);

            onFinished(res, function (err, res) {
                threads.releaseThreadId(threadId);
            });

            next();
        });
    };
};
